### PR TITLE
Yarn 1.x: Parse packages from yarn.lock

### PR DESCRIPTION
**Pull request by taylormadore**
This PR creates "Packages" for a Yarn 1.x project and all project dependencies. These "Packages" will later be resolved into SBOM Components in future PRs.

The yarn.lock file is the principal source for the fully resolved set of direct project dependencies and transitive dependencies (Yarn workspaces are an exception to this and must be parsed from package.json).

The pyarn library will be used to parse yarn.lock. There is a [related PR](https://github.com/containerbuildsystem/pyarn/pull/25) in that repository that must be merged and released first. It adds handling for additional styles of path-based version specifiers.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

File: tests/unit/package_managers/yarn_classic/test_resolver.py
Comment (a-ovchinnikov) [inline tests/unit/package_managers/yarn_classic/test_resolver.py:45]: What do you think about splitting the test in two: a positive and a negative test?

```python
@mark.parametrize("url",
 ["https://foo.com/bar.tgz", ...
 ],
)
def test__is_tarball_url_can_parse_correct_tarball_urls(url):
    assert _is_tarball_url(url)

@mark.parametrize("url",
 ["https://foo.com/bar.tar", ...
 ],
)
def test__is_tarball_url_rejects_incorrect_tarball_urls(url):
    assert not _is_tarball_url(url)
```

**Review by a-ovchinnikov — COMMENTED** [review]
Mostly LGTM, a couple of small nitpicks.

File: requirements-extras.txt
Comment (taylormadore) [reply requirements-extras.txt:664]: Yes exactly, the PR linked in the description needs to be merged first

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (eskultety) [inline cachi2/core/package_managers/yarn_classic/resolver.py:231]: minor nitpick (originating in an earlier commit) : overoptimization suggestion - do we need to create yet another intermediary list in the memory for something that is going to be iterated once only for SBOM components creation? IOW isn't `itertools.chain` a slightly better choice here instead of list concatenation?

File: cachi2/core/package_managers/yarn_classic/main.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/main.py]: Yikes, thanks for catching this. I thought I had fixed it already. Should be all set now

File: requirements-extras.txt
Comment (taylormadore) [reply requirements-extras.txt:349]: Moved to a separate commit

File: tests/integration/test_yarn_classic.py
Comment (taylormadore) [reply tests/integration/test_yarn_classic.py:116]: Do we prefer that? I don't have a strong opinion, but I did like the updates to be more visible/deliberate

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/resolver.py:198]: I pulled in the PackageJson class from the yarn module here. It should eventually be replaced by whatever the result of https://github.com/containerbuildsystem/cachi2/pull/705 is once merged

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/resolver.py:93]: Added a comment to the commit message

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/resolver.py:21]: Added the link to the yarn source

> I wonder what Yarn does if it encounters a project from gitea

It should work, it would just need to match one of the GIT_PATTERN_MATCHERS :man_shrugging:

File: tests/integration/test_yarn_classic.py
Comment (eskultety) [reply tests/integration/test_yarn_classic.py:116]: Every change to a test requires a corresponding change in the itegration suite runner.

> Do we prefer that?

We recently did and the last several PRs that we merged that included integration test changes already use branch names.

**Review by eskultety — COMMENTED** [review]
ACK, but we should IMO wait for #705 and merge the pyarn change in the meantime.

File: tests/integration/test_yarn_classic.py
Comment (a-ovchinnikov) [reply tests/integration/test_yarn_classic.py:116]: > Do we prefer that?

I would strongly recommend to in case we don't. It is so much easier to maintain tests with human-friendly refs than with hashes.

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (eskultety) [reply cachi2/core/package_managers/yarn_classic/resolver.py:198]: Uhm, apparently github failed to include my response in this thread when submitting the review, so repeating:

> I pulled in the PackageJson class from the yarn module here. It should eventually be replaced by whatever the result of https://github.com/containerbuildsystem/cachi2/pull/705 is once merged

Oh, but this means it's going to be super easy to forget to update this accordingly given there's already working logic. This creates a direct dependency on #705 then.

**Review by a-ovchinnikov — APPROVED** [review]

File: tests/integration/test_yarn_classic.py
Comment (taylormadore) [reply tests/integration/test_yarn_classic.py]: Sorry, github collapsed the comment and I forgot about it. Fixed

File: pyproject.toml
Comment (taylormadore) [reply pyproject.toml]: Updated to the released version

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/resolver.py]: Done

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (slimreaper35) [inline cachi2/core/package_managers/yarn_classic/resolver.py:49]: While working on generating SBOM I saw that there is an optional qualifier for `repository_url`:
https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs

The [default](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm) repository/registry for NPM is `https://registry.npmjs.org` so the second npm registry that we support is `"https://registry.yarnpkg.com` could/should be potentially included in the PURL.

Am I missing something ?

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn_classic/resolver.py:228]: non-blocking nitpick: it bothered me a little that we need to instantiate `_YarnClassicPackageFactory` twice. Looking at the `create_package_from_workspace` function, it seems to be pretty straightforward and the only instance data it needs is `source_dir.path`.

Do you think we could simply perform the `WorkspacePackage` instantiation within this function here, instead of having it as a `_YarnClassicPackageFactory` member function?

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/resolver.py]: Missed this, but you're right. We might want the resolved URL for registry packages when we get to the PURL generation. Added.

File: cachi2/core/package_managers/yarn_classic/resolver.py
Comment (taylormadore) [reply cachi2/core/package_managers/yarn_classic/resolver.py]: Sure, done :+1:

**Review by brunoapimentel — APPROVED** [review]

**Comment by slimreaper35** [comment]
nitpick: We should drop this unused piece of code:
https://github.com/containerbuildsystem/cachi2/blob/96534daf121cbdfd85156644d5f6f311eb25d98b/cachi2/core/package_managers/yarn_classic/main.py#L111

**Review by slimreaper35 — APPROVED** [review]

**Pull request merged**
